### PR TITLE
[ML] Add lazy parsing for DatafeedConfig:Aggs,Query

### PR DESCRIPTION
**Pull request by benwtrent**
This adds lazy parsing for DatafeedConfig Aggregations and Queries. 

When creating a NEW Datafeed, the STRICT parser is used, which eagerly parses both to catch any formatting errors. 

When an already stored config is read in, the LENIENT parser is used and simply pulls in the aggs and datafeed as maps. When serializing on the wire or to XContent, the maps are used directly. However, if the in/out streams are of an earlier version, they are fully parsed before sending over the wire. 

Right now the BWC checks are for `Version.CURRENT` as I am not sure if we want to put this in 6.x. Let me know what you think.

When a lazy parsing issue occurs, I am currently throwing a `UNPROCESSABLE_ENTITY` REST error. However, the argument could be made that they should just be `IllegalArgument` exceptions or a `XContentParseException`, need feedback :).

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java [hunk 1]
```diff
@@ -60,6 +64,45 @@
...
     private static final int TWENTY_MINS_SECONDS = 20 * SECONDS_IN_MINUTE;
     private static final int HALF_DAY_SECONDS = 12 * 60 * SECONDS_IN_MINUTE;
+    static final XContentObjectTransformer<QueryBuilder> QUERY_TRANSFORMER = XContentObjectTransformer.queryBuilderTransformer();
+    private static final BiFunction<Map<String, Object>, String, QueryBuilder> lazyQueryParser = (objectMap, id) -> {
+        try {
+            return QUERY_TRANSFORMER.fromMap(objectMap);
+        } catch (IOException | XContentParseException exception) {
+            // Certain thrown exceptions wrap up the real Illegal argument making it hard to determine cause for the user
+            if (exception.getCause() instanceof IllegalArgumentException) {
+                throw ExceptionsHelper.unprocessableEntityException(
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java]: This and other function handlers throw `ExceptionsHelper.unprocessableEntityException`. Should this be kept as such or move to `XContentParseException` or possibly `IllegalArgumentException` ? So as not to dictate the RESTful response code at this layer.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java [hunk 2]
```diff
@@ -564,14 +683,16 @@ public DatafeedConfig build() {
                 throw invalidOptionValue(TYPES.getPreferredName(), types);
             }
 
-            validateAggregations();
-            setDefaultChunkingConfig();
+            AggregatorFactories.Builder parsedAggs = lazyAggParser.apply(aggregations, id);
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java]: This is sort of tricky, when a user calls `GET` for a DatafeedConfig, `DatafeedConfig.Builder#build()` is called, I think...

If it is (still reading through to verify), I may have to figure out how to do these validations WITHOUT parsing the aggregations.

Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java]: ugh, yeah, verified by hijacking a yml test and forcing to always use lenient parsing:

```
"Test put datafeed with invalid query":
  - do:
      xpack.ml.put_datafeed:
        datafeed_id: test-datafeed-1
        body:  >
          {
            "job_id":"datafeeds-crud-1",
            "indexes":["index-foo"],
            "types":["type-bar"],
            "query":{"match_all_misspelled":"thing"}
            "aggregations": {"max":"invalid"}
          }
  - do:
      xpack.ml.get_datafeeds:
        datafeed_id: test-datafeed-1

  - match: { datafeeds.0.datafeed_id: "test-datafeed-1" }
  - match: { datafeeds.0.query.match_all_misspelled: "thing" }
  - match: { datafeeds.0.aggregations.max: "invalid" }
```

This fails :(. Gotta try and do the agg validations without parsing them

Comment (davidkyle) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java]: > If it is (still reading through to verify), I may have to figure out how to do these validations WITHOUT parsing the aggregations.

I would like the builders not to validate when `build()` is called instead they only need to validated when created or the datafeed is started. With the config in the index the current implementation validates every time the document is read.  #34899 is on my wishlist

Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java]: Thanks for the feedback y'all, I will get these concerns addressed soon.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java (hunk 1 above)
Comment (droberts195) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java]: I think that for BWC the aim should be to keep the exceptions for invalid syntax from a PUT datafeed the same as they were in 6.5.  So `unprocessableEntityException` (422) is almost certainly not right.

Please can you do an experiment and check what exception gets thrown if you do a PUT datafeed with an invalid query or aggregation in the datafeed using the master branch code.  Then change this code so the end user experience is identical when calling PUT datafeed.

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**